adapters/repos/db/lsmkv/roaringset/serialization.go

```diff
@@ -0,0 +1,194 @@
+package roaringset
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"math"
+
+	"github.com/dgraph-io/sroar"
+	"github.com/semi-technologies/weaviate/adapters/repos/db/lsmkv/segmentindex"
+)
+
+// A RoaringSet Segment Node stores one Key-Value pair (without its index) in
+// the LSM Segment.  It uses a single []byte internally. As a result there is
+// no decode step required at runtime. Instead you can use
+//
+//   - [*SegmentNode.Additions]
+//   - [*SegmentNode.AdditionsWithCopy]
+//   - [*SegmentNode.Deletions]
+//   - [*SegmentNode.DeletionsWithCopy]
+//   - [*SegmentNode.PrimaryKey]
+//
+// to access the contents. Those helpers in turn do not require a decoding
+// step. The accessor methods that return Roaring Bitmaps only point to
+// existing memory (methods without WithCopy suffix), or in the worst case copy
+// one byte slice (methods with WithCopy suffix).
+//
+// This makes the SegmentNode very fast to access at query time, even when it
+// contains a large amount of data.
+//
+// The internal structure of the data is:
+//
+//	byte begin-start    | description
+//	--------------------|-----------------------------------------------------
+//	0-8                 | uint64 indicating the total length of the node,
+//	                    | this is used in cursors to identify the next node.
+//	8-16                | uint64 length indicator for additions sraor bm -> x
+//	16-(x+16)           | additions bitmap
+//	(x+16)-(x+24)       | uint64 length indicator for deletions sroar bm -> y
+//	(x+24)-(x+y+24)     | deletions bitmap
+//	(x+y+24)-(x+y+28)   | uint32 length indicator for primary key length -> z
+//	(x+y+28)-(x+y+z+28) | primary key
+type SegmentNode struct {
+	// Offset is not persisted in the node itself, but it is respected when
+	// building a [segmentindex.Key] using [*SegmentNode.KeyIndexAndWriteTo].
+	// multiple nodes
+	Offset int
```

Comment:
can `Offset` be an argument of `KeyIndexAndWriteTo` method as it is used only there?

Reply:
Good point. My initial motivation was to keep the signature of `KeyIndexAndWriteTo` consistent with other usages of it. But this is never abstracted through an interface, so this consistency has very little value. And I agree that it's much more explicit to pass it when it's used.

Reply:
I've seen other `KeyIndexAndWriteTo` implementations, but couldn't find the interface for them, hence my suggestion for change.